configure.ac

```diff
@@ -122,6 +143,23 @@ AC_CHECK_FUNCS([memset strchr strdup strerror strrchr])
 
 AC_SUBST(NATIVE_DIR)
 
+# Version
+AC_SUBST(ADFLIB_MAJOR_VERSION, [adflib_version_major])
+AC_SUBST(ADFLIB_MINOR_VERSION, [adflib_version_minor])
+AC_SUBST(ADFLIB_MICRO_VERSION, [adflib_version_micro])
+
+# Libtool version info
+# ( https://www.gnu.org/software/libtool/manual/html_node/Updating-version-info.html )
+#AC_SUBST(LT_RELEASE, [adflib_version_major.adflib_version_minor])
+#LT_CURRENT=adflib_version_major
+LT_CURRENT=`expr 100 '*' adflib_version_major + adflib_version_minor`
+LT_REVISION=adflib_version_micro
+#LT_AGE=adflib_version_minor
+LT_AGE=0
+AC_SUBST(LT_CURRENT)
+AC_SUBST(LT_REVISION)
+AC_SUBST(LT_AGE)
```

Comment:
Yes, this is something where some decision should be made, I kind of expected discussion about it. Good :-)

Some way of doing this has to be chosen. Proposing this, I considered it a reasonably simple yet working solution.

I know what libtool doc suggests (I put the link in the config...). However, I had a brief look how misc. libraries deal with it - and they _usually_ do something like I proposed here. We can have a rule that maps the library versioning to libtool/soname in some reasonable way, assuming _strictly_ following versioning rules.

Consider versioning of ADFlib as follows:

    `version_major.version_minor.version_micro`

where:
- `version major` increase is like a milestone (big library revision), clearly API/ABI breaking change
- `version minor` increase can be
  - an API/ABI breaking change or
  - a non-breaking change (an extension), in which case `age` should be increased as well(!), (but of course it must be reset if an API/ABI breaking change occurs)
- `version micro` can be only just a revision (an internal change, no changes in the API)

That can be mapped to libtool's version:

    `version_major * 100 + version_minor:version_micro:age`

where `age` is specified _separately_ in `configure.ac` (see the piece you put above)

Assuming that the `version_minor` will be 0-99 (a rather safe assumption), this seems to do exactly what is expected by libtool/soname rules.

Now:
- the version was not changed in the repo yet, so for sure it will not be `7:7:0`... - if you put the ADFlib version `0.8.0` (as discussed) to the mapping rule, you will get `8:0:0` for libtool and `libadf.so.8` for soname.
- I have just built version 0.7.12 (the last released), its soname is `libadf.so.0`

So, if I understand well, `.8` _will_ indicate binary incompatibility vs. `.0` of the latest released version (0.7.12).

I understand, you may like to have `.1` instead, but from practical point of view, there is no difference.

In the future changes, in my understanding, this would also work properly.

Note that merging this does not imply that we have to release it this way, we can change it in any way agreed.

(Eventually, we can move this discussion to an issue, and work out clearly what will be the policy for versioning).

Reply:
You agree a package version is not enough for determining library version, the extra variable "age" is _needed_.

Introducing a scheme for tying package version to `-version-info` , when coupled with a motivated person who knows this scheme (they possibly even invented the scheme) and is always available to do the maintenance of the "age" variable that works correctly with this special scheme... sure, technically, that will work.

But what is simpler and even easier: no special scheme.

* Have a package version number. Update it with whatever rules you prefer today, including semver or almost-semver. You can change your mind in future and it doesn't break binary compatibility as a side-effect.
* Have a library version number. Update it with the libtool rules.

I would be happy to write a `bump-version` script. In fact, here is one I've just written. It could easily be extended to read the current version differentl, and write the new version to more than one file.

```bash
#!/bin/sh
# Usage: util/bump-version <no-api-changes|compatible-api-changes|breaking-api-changes>

# assuming we're $basedir/util/bump-version, find $basedir
dir=`dirname "$0"`; basedir=`cd "$dir"/.. && pwd`

# extract current package and library version from configure.ac
old_pkg_version=`grep ^ADFLIB_VERSION= "$basedir/configure.ac" | sed 's/^ADFLIB_VERSION=//'`
old_lt_version=`grep ^ADFLIB_LT_VERSION= "$basedir/configure.ac" | sed 's/^ADFLIB_LT_VERSION=//'`

# split versions into their components
pkg_major=`echo $old_pkg_version | sed 's/\..*//'`
pkg_minor=`echo $old_pkg_version | sed -e 's/[^.]*\.//' -e 's/\..*//'`
pkg_patch=`echo $old_pkg_version | sed -e 's/.*\.//' -e 's/[^0-9]//g'`
lt_current=`echo $old_lt_version | sed 's/:.*//'`
lt_revision=`echo $old_lt_version | sed -e 's/[^:]*://' -e 's/:.*//'`
lt_age=`echo $old_lt_version | sed -e 's/.*://'`

# update versions
case "$1" in
    no-api-changes)
       pkg_patch=`expr $pkg_patch + 1`
       lt_revision=`expr $lt_revision + 1`
    ;;
    compatible-api-changes)
       pkg_minor=`expr $pkg_minor + 1`
       pkg_patch=0
       lt_current=`expr $lt_current + 1`
       lt_revision=0
       lt_age=`expr $lt_age + 1`
    ;;
    breaking-api-changes)
       # note that adflib package versioning _IS NOT SEMVER_
       # semver would bump pkg_major for breaking changes, adflib chooses not to
       pkg_minor=`expr $pkg_minor + 1`
       pkg_patch=0
       lt_current=`expr $lt_current + 1`
       lt_revision=0
       lt_age=0
    ;;
    *)
    echo >&2 "Usage: $0 <no-api-changes|compatible-api-changes|breaking-api-changes>"
    exit 1
    ;;
esac

new_pkg_version="$pkg_major.$pkg_minor.$pkg_patch"
new_lt_version="$lt_current:$lt_revision:$lt_age"

echo "Bumping package version $old_pkg_version -> $new_pkg_version"
echo "Bumping library version $old_lt_version -> $new_lt_version"

# update version numbers in configure.ac
sed -i.bak -e "s/^ADFLIB_VERSION=.*/ADFLIB_VERSION=$new_pkg_version/" \
           -e "s/^ADFLIB_LT_VERSION=.*/ADFLIB_LT_VERSION=$new_lt_version/" \
    "$basedir/configure.ac"
```

Reply:
I think both ways have its advantages and disadvantages, eg.

* "tied" versioning 
  `+` the lib version relates to soname - checking the packages etc. not needed
  `+-` a _very_ careful bumping is needed
  `-` not quite follow the libtool's guidelines
- "untied" versioning
  `+` truly comply with libtool's guidelines, can nicely progress with libtool scheme (increasing by 1 when needed)
  `+-` no relation between soname and the library/package version
  `-` must make changes in 2 independent schemes with (possibly) different rules

Simplicity of both - that is up to a debate... Most libraries that I checked do the "tied" one, so it is not that uncommon and inherently bad solution, if done carefully.

But I tend to agree with you that the "untied" way might be cleaner (though personally I do not like these completely unrelated versions in sonames). Having the script you prepared, it is fine for me this way. So far it was hardcoded, so any way we improve this will be OK.

However, maybe it would be worth to note there is not only autotools' way of building and currently there are a few more things regarding versioning:
- deb packaging - which requires the same bumping as lib version (this is somewhat "unofficial" packaging, the distros may do this their way, anyway...)
- cmake builds have no versioning for the binary (soname/dll) so far; these build configs are done mostly for wider usage and testing, to facilitate builds in misc. environments (eg. on windows) - but doing this really "properly" may require some more work...

For these, we will have several ways: leave it as it is (manual for debs, nothing for CMake builds), extend your script or create dedicated ones (one for deb, one for CMake). For me, we can leave it as it is for now (manual update of deb, some manual changes are required anyway... and no versioning of soname/dll for CMake) and improve it later. Spending too much time on that at this moment, when the lib is in rather unstable state, is maybe a bit early (better leave building the binary without version inside that do it badly...).

At the moment, versioning the binary seems more important for autotools builds - as it may really produce something that might be included in some distros at some point. Though who knows what people will use...

This may require further discussion (maybe in a dedicated issue) to eventually cover (or at least have a plan) how to deal with all cases (autotools alone, autotools+deb, cmake for Linux/Mac/Windows, ... ???).

For now, to keep things safely compatible with past and any future (even if "tied"), maybe we (ie. you ;-) could apply what you propose (after we merge this). And then, eventually, the rest can be improved as well. Would this be OK?